### DISCUSSION TITLE
Setup Questions/Configuration

### DISCUSSION THREAD

**Discussion by challgren** [3DCoded/3dchameleon-klipper#3] [General]
I was reading the docs and I can't understand how to hook up the limit switch to the relay. Can you detail this or provide a photo?

Thanks!

**Comment by 3DCoded** [comment]
Sure! I have added a document explaining how to hook up the limit switch, Chameleon, and RPi to the Relay. Note that this uses a Relay shield as I don't have much experience with normal relays. Here is the link: [docs/wiring.md](https://github.com/3DCoded/3dchameleon-klipper/blob/main/docs/wiring.md)

If this answers your question, please close this issue. If not, please tell me what I can do to help answer it more clearly.

**Comment by challgren** [comment]
Thank you very much! I don't have the kit just yet but I've been researching on how to use it with klipper and my first thing was to get rid of the physical button but I didn't know how people achieved that. But thank you that clears my issue up and should help others on their journey.

**Comment by challgren** [comment]
After reading the 3d chameleon forums you can directly hook up to the raspberry pi by inverting the pin and connecting green to your gpio and black to a ground. 2.0v+ will trigger the chameleon. But thank you for the guide.

**Comment by 3DCoded** [comment]
Thank you. I didn't see that. I will experiment with that to see if any changes would need to be made to the plugin/documentation.

**Comment by challgren** [comment]
The biggest would be inversion of the pin. I was using your script and the dependency on the filament sensor is kind of a deal breaker right now. Also having a lot of config file parsing issues.

**Comment by 3DCoded** [comment]
I have updated the documentation (the Wiki) on how to invert the pin, but the filament sensor is needed for reliability reasons (the entire plugin is built around a filament sensor). What kind of config file parsing issues are you running into? Can you post your error messages here?

**Reply by challgren** [reply]
As of now none but I did update to the latest version of your plugin. I did copy the config from the wiki and I was getting issues parsing load_time

**Reply by 3DCoded** [reply]
Oh... The code in the backend was trying to parse the `load_time` as an integer (whole number), not a float (decimal number). I just pushed the fix. Thank you for pointing that out.

**Comment by 3DCoded** [comment]
I just added experimental support for disabling the filament sensor, which can be used by setting `filament_sensor_name: disabled`.

**Reply by challgren** [reply]
Awesome I'll test it out! I will probably do a PR for an install script so people can update via moonraker

**Reply by 3DCoded** [reply]
I'm in the middle of a long print, so I'm not able to test this now, but I just added a simple experimental installation and update script. What do you think about them? This is my first time writing a klippy extra, so I don't know if there's an already agreed method of doing this.

**Reply by challgren** [reply]
I would say do a ln -s in klipper/extras instead of leaving a dirty file. I kind of looked at the way klipper-led_effects repo is setup.

**Reply by 3DCoded** [reply]
Good idea. I implemented that into the `update.sh`, as it's the only one that messes with `klippy/extras`. I kept the `install.sh` as is, so that `3dchameleon.cfg` is still easy to edit.

**Comment by challgren** [comment]
The only other thing now I can think the plugin would need would be the ability to load filament into the mmu. I'm starting from fresh and I'm having issues with initial loading and my path being possibly too long or the filament sensor is causing a jam on the pathway

**Reply by challgren** [reply]
So right now I was able to figure out the correct loading sequence. Which from my understanding is unplug the control box, move around the gear box and feed each filament until about 1" from the Y tubing. I was able to successfully complete that. With a little help from https://github.com/alexpthe1/my3DChameleon I was able to use some macros to start understanding the Chameleon more. After I load I know the next step is to home without filament movement. I guess I need to figure a macro how to load the current filament into the hotend extruder. It would be nice to have a macro/function to be able to move filament +/- by seconds/mms

**Reply by 3DCoded** [reply]
To home without filament movement, this should work:
`PULSE_CHAMELEON PULSES=7`

I can work on a macro/function to move filament forwards (towards the extruder). This can technically be accomplished by:
```
PULSE_CHAMELEON PULSES=7
PULSE_CHAMELEON PULSES={tool}
PRESS_CHAMELEON DURATION={duration}
```
To move the filament backwards (away from the extruder) would complicate the algorithm, and I think the solution you described above would be best.
